File: Angular/src/Dominio/Gustos/Gustos.ts
import { Vehiculo } from "../vehiculo/vehiculo"

export type GustoJson = {

    type:string;
    marcaF: string;
    listaDeGustos: Gusto[]

}

export class Gusto{

    readonly type:string = ""

    condicion(auto: Vehiculo): boolean{
        return false
    }

    getTipo(): string[] {
        return [this.type]
    }

    static fromJson(gustoJson:GustoJson):Gusto{
    //Se convierten todos en combinados asi es mas facil manejarlos en el front

        switch(gustoJson.type){

            case "Neofilo" : {
                return new Combinado([new Neofilo()])
            }
            case "Supersticioso" : {
                return new Combinado([new Supersticioso()])
            }
            case "Caprichoso" : {
                return new Combinado([new Caprichoso()])
            }
            case "Selectivo" : {
                return new Combinado([new Selectivo(gustoJson.marcaF)])
            }
            case "SinLimite" : {
                return new Combinado([new SinLimite()])
            }
            case "Combinado" : {
                return new Combinado(gustoJson.listaDeGustos)
            }
        }
        throw new Error("Tipo Invalido")
    }
 }

export class Neofilo extends Gusto {

    override type: string = "Neofilo";

    override condicion(auto: Vehiculo): boolean {
        return auto.antiguedad() < 2
    }
}

export class Supersticioso extends Gusto {

    override type: string = "Supersticioso";

    override condicion(auto: Vehiculo): boolean {
        return auto.anioDeFabricacion.getFullYear() % 2 == 0

    }
}
export class Caprichoso extends Gusto {

    override type: string = "Caprichoso";

    override condicion(auto: Vehiculo): boolean {
        return auto.modelo.charAt(0).toUpperCase() ==  auto.marca.charAt(0).toUpperCase()
    }

}
export class Selectivo extends Gusto {

    marcaF:string

    override type: string = "Selectivo"

    constructor(_marcaF:string){
        super()
        this.marcaF = _marcaF
    }
    
    override condicion(auto: Vehiculo): boolean {
        return auto.marca == this.marcaF
    }


}
export class SinLimite extends Gusto {

    override type: string = "SinLimite"

    override condicion(auto: Vehiculo): boolean {
         return auto.kilometrajeLibre
    }

}
export class Combinado extends Gusto {

    listaDeGustos:Array<Gusto>
    override type: string = "Combinado"

    constructor(_listaDeGustos:Array<Gusto>){
        super()
        this.listaDeGustos = _listaDeGustos
    }
    
    override getTipo():string[]{
        return this.listaDeGustos.map( gusto => gusto.type)
    }

    agregarGusto(gusto: Gusto){
        this.listaDeGustos.push(gusto)
    }

    quitarGusto(gusto: Gusto){
        const index = this.listaDeGustos.indexOf(gusto, 0);
        if (index > -1) {
            this.listaDeGustos.splice(index, 1);
        }
    }

    getGustoEnLista(tipo:string):Gusto{
        let gusto = this.listaDeGustos.find( it => it.type == tipo)
        if(gusto == null){
            throw("El tipo de gusto no se encuentra en la lista de gustos")
        }
        return gusto
    }

    override condicion(auto: Vehiculo): boolean {
        return this.listaDeGustos.every((gusto) =>{ return gusto.condicion(auto)})
    }

}

File: hola-mundo-backend-2022-grupo-7/src/main/kotlin/trabajoPracticoAlgoritmos2/Gustos.kt
package trabajoPracticoAlgoritmos2

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo

@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.PROPERTY,
    property = "type")
@JsonSubTypes(
    JsonSubTypes.Type(value = Neofilo::class, name = "Neofilo"),
    JsonSubTypes.Type(value = Supersticioso::class, name = "Supersticioso"),
    JsonSubTypes.Type(value = Caprichoso::class, name = "Caprichoso"),
    JsonSubTypes.Type(value = Selectivo::class, name = "Selectivo"),
    JsonSubTypes.Type(value = SinLimite::class, name = "SinLimite"),
    JsonSubTypes.Type(value = Combinado::class, name = "Combinado")
)
open class Gustos{
    open fun condicion(auto: Vehiculo): Boolean {
        return false
    }
}

class Neofilo: Gustos() {
    override fun condicion(auto: Vehiculo): Boolean {
        return auto.antiguedad() < 2
    }

}
class Supersticioso: Gustos() {
    override fun condicion(auto: Vehiculo): Boolean {
        return auto.anioDeFabricacion.year % 2 == 0

    }
}
class Caprichoso: Gustos() {
    override fun condicion(auto: Vehiculo): Boolean {
        return auto.modelo.first().uppercase() ==  auto.marca.first().uppercase()
    }

}
class Selectivo(
    var marcaF:String
): Gustos() {

    override fun condicion(auto: Vehiculo): Boolean {
        return auto.marca == marcaF
    }


}
class SinLimite: Gustos() {
    override fun condicion(auto: Vehiculo): Boolean {
         return auto.kilometrajeLibre
    }

}
class Combinado(val listaDeGustos: MutableList<Gustos>): Gustos() {

    fun agregarGusto(gusto: Gustos){
        listaDeGustos.add(gusto)
    }

    override fun condicion(auto: Vehiculo): Boolean {
        return listaDeGustos.all { it.condicion(auto) }
    }

}

